css/cssom/caretRangeFromPoint.tentative.html

```diff
@@ -0,0 +1,121 @@
+</style>
+<div id="div">abc</div>
+<div id="shadow">
+  <template shadowrootmode="open">
+    <div>def</div>
+  </template>
+</div>
+
+<canvas id="canvas" width="500" height="500"></canvas>
+
+<script>
+  test(() => {
+    let range = document.caretRangeFromPoint();
+    assert_true(range instanceof Range);
+    assert_equals(range.startOffset, 0);
+    assert_equals(range.endOffset, 0);
+  }, "document.caretRangeFromPoint() (no supplied coordinates) returns Range with 0 0 values");
+
+  test(() => {
+    const doc = document.implementation.createHTMLDocument("");
+    let range = doc.caretRangeFromPoint(0, 0);
+    assert_equals(range, null);
+  }, "document.caretRangeFromPoint() should return null for a document with no viewport");
+
+  test(() => {
+    assert_equals(document.caretRangeFromPoint(-5, 5), null);
+    assert_equals(document.caretRangeFromPoint(5, -5), null);
+    assert_equals(
+      document.caretRangeFromPoint(document.documentElement.clientWidth * 2, 5),
+      null,
+    );
+    assert_equals(
+      document.caretRangeFromPoint(
+        5,
+        document.documentElement.clientHeight * 2,
+      ),
+      null,
+    );
+  }, "document.caretRangeFromPoint() should return null if given coordinates outside of the viewport");
+
+  test(() => {
+    const rect = div.getBoundingClientRect();
+    const characterWidth = rect.width / div.textContent.length;
+    const characterIndex = 2;
+    const x = rect.left + characterWidth * characterIndex;
+    const y = rect.top + rect.height / 2;
+    const range = document.caretRangeFromPoint(x, y);
+    assert_true(range instanceof Range);
+    assert_equals(range.startContainer, div.childNodes[0]);
+    assert_equals(range.endContainer, div.childNodes[0]);
+    assert_equals(range.startOffset, characterIndex);
+    assert_equals(range.endOffset, characterIndex);
+    assert_true(range.collapsed);
+    range.setEnd(div.childNodes[0], characterIndex + 1);
+    assert_false(range.collapsed);
+    assert_equals(range.toString(), "c");
+    range.setStart(div.childNodes[0], characterIndex - 1);
+    assert_false(range.collapsed);
+    assert_equals(range.toString(), "bc");
+    range.setStart(div.childNodes[0], 0);
+    assert_false(range.collapsed);
+    assert_equals(range.toString(), "abc");
+  }, "document.caretRangeFromPoint() should return a Range at the specified location");
+
+  test(() => {
+    const rect = div.getBoundingClientRect();
+    const characterWidth = rect.width / div.textContent.length;
+    const characterIndex = 2;
+    const x = rect.left + characterWidth * characterIndex;
+    const y = rect.top + rect.height / 2;
+    const range = document.caretRangeFromPoint(x, y);
+    let rangeRect = range.getBoundingClientRect();
+    assert_approx_equals(rangeRect.x, x, 1);
+    assert_approx_equals(rangeRect.y, rect.height / 2, 3);
+    assert_approx_equals(rangeRect.height, rect.height, 1);
+    assert_equals(rangeRect.width, 0, "Caret range should be collapsed");
```

Comment:
This line fails in Blink, with `width = 0.015625`. Should this also use `assert_approx_equals`?